### PR TITLE
fuzz: make h1_capture_fuzz_test more robust.

### PR DESCRIPTION
**Pull request by htuch**
Eliminate some more false positives. This addresses oss-fuzz issue 8341.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by alyssawilk** [comment]
@htuch what's the status on this one?

**Comment by htuch** [comment]
I'll take a look sometime in the next couple of days, moving the fuzzing along is a relatively low priority background thread for me right now.

**Comment by mattklein123** [comment]
@htuch can we potentially get this merged? I think this might fix some flakes we are seeing?

**Comment by htuch** [comment]
@mattklein123 ignore this for now, I'm debugging.

**Comment by htuch** [comment]
Looks like there are multiple race conditions that get exposed as flakes when I do `--runs_per_test=1000 -c dbg --config=clang-asan`. They mostly relate to how we do connection object shutdown and deferred delete in the `FakeUpstream`, the interplay with possible outstanding dispatcher events referring to this state and the `FakeRawConnection` being used by the test driver. I think some are latent issues in the test framework itself, so fixing them will help with stability.

I'm going to leave this as-is while I weed out the remaining issues.

**Comment by htuch** [comment]
I should also add that I've now verified this with `--config=clang-asan --runs_per_test=1000 -c dbg`, which is where I was hitting a large number of corner cases in `h1_capture_fuzz_test` previously.

**Comment by mattklein123** [comment]
@htuch just noticed the CI failures. Can you take a look and then I will review?

**Comment by alyssawilk** [comment]
Yeah, I'll take a look as well once tsan is happy

File: test/integration/fake_upstream.h [hunk 1]
```diff
@@ -130,6 +131,96 @@ class FakeStream : public Http::StreamDecoder,
...
+  // This provides direct access to the underlying connection, but only to const methods.
+  // Stateful connection related methods should happen on the connection's dispatcher via
+  // executeOnDispatcher.
+  // TODO(htuch): This seems a sketchy pattern; even if we're using const methods, there may be
+  // thread safety violations when crossing between the test thread and FakeUpstream thread.
+  Network::Connection& connection() const { return connection_; }
+
+  // Execute some function on the connection's dispatcher. This involves a cross-thread post and
+  // wait-for-completion. If the connection is disconnected, either prior to post or when the
+  // dispatcher schedules the callback, we silently ignore, unless panic_on_unexpected_disconnect_
```
Comment (alyssawilk) [inline test/integration/fake_upstream.h]: unless allow_unexpected_disconnects_ is false?

File: test/integration/fake_upstream.h [hunk 2]
```diff
@@ -130,6 +131,96 @@ class FakeStream : public Http::StreamDecoder,
...
+    Thread::LockGuard lock(lock_);
+    return !disconnected_;
+  }
+
+  // This provides direct access to the underlying connection, but only to const methods.
+  // Stateful connection related methods should happen on the connection's dispatcher via
+  // executeOnDispatcher.
+  // TODO(htuch): This seems a sketchy pattern; even if we're using const methods, there may be
+  // thread safety violations when crossing between the test thread and FakeUpstream thread.
+  Network::Connection& connection() const { return connection_; }
```
Comment (alyssawilk) [inline test/integration/fake_upstream.h]: Can this be made const, or is this comment about const methods wishful thinking?

Comment (htuch) [reply test/integration/fake_upstream.h]: Wishful thinking; even a dispatcher post via the connection is non-const :/

Comment (mattklein123) [reply test/integration/fake_upstream.h]: Yeah agreed, not sure if this will work in all cases, but we can see what else breaks and fix it.

**Comment by htuch** [comment]
@alyssawilk 's suggestion to run internal TSAN was fruitful, squashed some more races. It now passes with `//test/integration/... --runs_per_test=200 --config=tsan`, so this seems reasonable. @mattklein123 can you rubber stamp again, assuming it passes on CircleCI with this update? Thank!

**Pull request merged**